builtin/logical/azuresql/backend.go

```diff
@@ -0,0 +1,194 @@
+	server      string
+	lock        sync.Mutex
+}
+
+func (b *backend) AzureClient(s logical.Storage) (*azsql.SQLDatabaseClient, error) {
+	b.lock.Lock()
+	defer b.lock.Unlock()
+
+	// If we already have a client, just return it!
+	if b.azureClient != nil {
+		return b.azureClient, nil
+	}
+
+	// Otherwise, attempt to make connection
+	entry, err := s.Get("config/subscription")
+	if err != nil {
+		return nil, err
+	}
+	if entry == nil {
+		return nil, fmt.Errorf("configure the Azure client connection with config/subscription first")
+	}
+
+	var azureConfig subscriptionConfig
+	if err := entry.DecodeJSON(&azureConfig); err != nil {
+		return nil, err
+	}
+	// Use the Azure Go SDK
+	var client management.Client
+	if len(azureConfig.PublishSettings) > 0 {
+		client, err = management.ClientFromPublishSettingsFile(azureConfig.PublishSettings, azureConfig.SubscriptionID)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		cert, err := ioutil.ReadFile(azureConfig.ManagementCert)
+		if err != nil {
+			return nil, err
+		}
+		client, err = management.NewClient(azureConfig.SubscriptionID, cert)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	//Test out the client
+	azClient := azsql.NewClient(client)
+	_, err = azClient.ListFirewallRules(azureConfig.Server)
+	if err != nil {
+		return nil, err
+	}
+
+	b.azureClient = &azClient
+	b.server = azureConfig.Server
+	return b.azureClient, nil
+}
+
+// DB returns the default database connection.
+func (b *backend) DB(s logical.Storage) (*sql.DB, error) {
+	b.lock.Lock()
+	defer b.lock.Unlock()
+
+	// If we already have a DB, we got it!
+	if b.db != nil {
+		return b.db, nil
+	}
+
+	// Otherwise, attempt to make connection
+	entry, err := s.Get("config/connection")
+	if err != nil {
+		return nil, err
+	}
+	if entry == nil {
+		return nil, fmt.Errorf("configure the DB connection with config/connection first")
+	}
+
+	var connConfig connectionConfig
+	if err := entry.DecodeJSON(&connConfig); err != nil {
+		return nil, err
+	}
+	connString := connConfig.ConnectionString
+
+	db, err := sql.Open("mssql", connString)
+	if err != nil {
+		return nil, err
+	}
+
+	// Set some connection pool settings. We don't need much of this,
+	// since the request rate shouldn't be high.
+	db.SetMaxOpenConns(connConfig.MaxOpenConnections)
+
+	//Testing out the connection
+	stmt, err := db.Prepare("SELECT db_name();")
+	if err != nil {
+		return nil, err
+	}
+	defer stmt.Close()
+
+	err = stmt.QueryRow().Scan(&b.defaultDb)
+	if err != nil {
+		return nil, err
+	}
+
+	b.db = db
+	return b.db, nil
+}
+
+// ResetDB forces a connection next time DB() is called.
+func (b *backend) ResetDB() {
+	b.lock.Lock()
+	defer b.lock.Unlock()
+
+	if b.db != nil {
+		b.db.Close()
+	}
+
+	b.db = nil
+}
+
+// LeaseConfig returns the lease configuration
+func (b *backend) LeaseConfig(s logical.Storage) (*configLease, error) {
+	entry, err := s.Get("config/lease")
+	if err != nil {
+		return nil, err
+	}
+	if entry == nil {
+		return nil, nil
```

Comment:
returning nil nil is weird since you are likely to run into something like if err != nil {} configLease.Foo throwing a panic.